src/Compilers/CSharp/Test/Semantic/Semantics/WarningVersionTests.cs

```diff
@@ -47,5 +49,244 @@ struct S
+            var comp1 = CreateCompilation(source, options: TestOptions.DebugModule);
+            var moduleReference = comp1.EmitToImageReference();
+
+            var source2 =
+@"public struct Program
+{
+    public Struct Field;
+    Program(int dummy)
+    {
+        Program p2 = this;
+        this.Field = default;
+    }
+}";
+            CreateCompilation(source2, references: new MetadataReference[] { moduleReference }).VerifyDiagnostics(
+                );
+            CreateCompilation(source2, references: new MetadataReference[] { moduleReference }, options: TestOptions.ReleaseDll.WithWarningLevel(5)).VerifyDiagnostics(
+                // (6,22): warning CS8827: The 'this' object cannot be used before all of its fields have been assigned
+                //         Program p2 = this;
+                Diagnostic(ErrorCode.WRN_UseDefViolationThis, "this").WithArguments("this").WithLocation(6, 22)
+                );
+        }
+
+        [Fact]
+        public void WRN_UseDefViolationOut()
+        {
+            var source = @"
+public struct Struct
+{
+    private string data;
+}
+";
+            var comp1 = CreateCompilation(source, options: TestOptions.DebugModule);
+            var moduleReference = comp1.EmitToImageReference();
+
+            var source2 =
+@"class Program
+{
+    public static void M(out Struct r1)
+    {
+        var r2 = r1;
+        r1 = default;
+    }
+}";
+            CreateCompilation(source2, references: new MetadataReference[] { moduleReference }).VerifyDiagnostics(
+                );
+            CreateCompilation(source2, references: new MetadataReference[] { moduleReference }, options: TestOptions.ReleaseDll.WithWarningLevel(5)).VerifyDiagnostics(
+                // (5,18): warning CS8828: Use of unassigned out parameter 'r1'
+                //         var r2 = r1;
+                Diagnostic(ErrorCode.WRN_UseDefViolationOut, "r1").WithArguments("r1").WithLocation(5, 18)
+                );
+        }
+
+        [Fact]
+        public void WRN_UseDefViolation()
+        {
+            var source = @"
+public struct Struct
+{
+    private string data;
+}
+";
+            var comp1 = CreateCompilation(source, options: TestOptions.DebugModule);
+            var moduleReference = comp1.EmitToImageReference();
+
+            var source2 =
+@"class Program
+{
+    public static void Main()
+    {
+        Struct r1;
+        var r2 = r1;
+    }
+}";
+            CreateCompilation(source2, references: new MetadataReference[] { moduleReference }).VerifyDiagnostics(
+                );
+            CreateCompilation(source2, references: new MetadataReference[] { moduleReference }, options: TestOptions.ReleaseDll.WithWarningLevel(5)).VerifyDiagnostics(
+                // (6,18): warning CS8829: Use of unassigned local variable 'r1'
+                //         var r2 = r1;
+                Diagnostic(ErrorCode.WRN_UseDefViolation, "r1").WithArguments("r1").WithLocation(6, 18)
+                );
+        }
```

Comment:
Is there a follow-up issue for the SkipLocalsInit scenario?

Reply:
I think the conclusion was that we don't need any particular behavior specific to the SkipLocalsInit scenario because the runtime clears reference fields.

---
In reply to: [459100259](https://github.com/dotnet/roslyn/pull/46221#discussion_r459100259) [](ancestors = 459100259)

Reply:
This is not the decision LDM made: [https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-07-20.md#private-fields-in-structs-with-skiplocalsinit](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-07-20.md#private-fields-in-structs-with-skiplocalsinit).

---
In reply to: [459150046](https://github.com/dotnet/roslyn/pull/46221#discussion_r459150046) [](ancestors = 459150046,459100259)

Reply:
The basis for that decision was the mistaken belief that the runtime does not clear reference fields with skiplocalsinit. In fact the runtime does clear them. If you think it necessary I will confirm with the LDM.

---
In reply to: [463275116](https://github.com/dotnet/roslyn/pull/46221#discussion_r463275116) [](ancestors = 463275116,459150046,459100259)

Reply:
Please do.

---
In reply to: [463308924](https://github.com/dotnet/roslyn/pull/46221#discussion_r463308924) [](ancestors = 463308924,463275116,459150046,459100259)

Reply:
That discussion is still ongoing so I have recorded [https://github.com/dotnet/roslyn/issues/46488](https://github.com/dotnet/roslyn/issues/46488) to follow up on it separately.

---
In reply to: [463324225](https://github.com/dotnet/roslyn/pull/46221#discussion_r463324225) [](ancestors = 463324225,463308924,463275116,459150046,459100259)